servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/CapturedContextProvider.java

```diff
@@ -17,53 +17,68 @@
+ *     }
+ *
+ *     private void setMyState(MyState myState) {
+ *         // set the context state in the local environment
+ *     }
+ *
+ *     CapturedContext captureContext(CapturedContext underlying) {
+ *          return new CapturedContextImpl(delegate, getMyState());
+ *     }
+ *
+ *     CapturedContext captureContextCopy(CapturedContext underlying) {
+ *          return new CapturedContextImpl(delegate, getMyState().copy());
+ *     }
+ * }</pre>
+ *
+ * <b>Note:</b> If the MyState type is immutable then there is no distinction between
+ * captureContext(..) and captureContextCopy(..)
  */
-@FunctionalInterface
+
 public interface CapturedContextProvider {
 
     /**
-     * Capture existing context in preparation for an asynchronous thread jump.
-     *
-     * If you want to capture any external state you can create a wrapper {@link CapturedContext} to add additional
-     * state capturing to the context pathway. This state can then be restored by wrapping the {@link CapturedContext}
-     * with the additional functionality to restore and finally revert the context state.
-     * <p>
-     * An example provider may be implemented as follows:
-     * <pre>{@code
-     *     private class CapturedContextImpl implements CapturedContext {
-     *         private final CapturedContext delegate;
-     *         private final String state;
-     *
-     *         public Scope attachContext() {
-     *             String old = getMyString();
-     *             setMyString(state);
-     *             Scope outer = delegate.attachContext();
-     *             return () -> {
-     *                 outer.close();
-     *                 setMyString(old);
-     *             };
-     *         }
-     *
-     *         public ContextMap captured() {
-     *             return delegate.captured();
-     *         }
-     *     }
-     *
-     *     private String getMyString() {
-     *         // capture context state from the local environment
-     *     }
-     *
-     *     private void setMyString(String string) {
-     *         // set the context state in the local environment
-     *     }
-     *
-     *     CapturedContext captureContext(CapturedContext underlying) {
-     *          return new CapturedContextImpl(delegate, getMyString());
-     *     }
-     * }</pre>
+     * Capture a reference to existing context in preparation for an asynchronous thread jump.
      * @param underlying additional context that <i><b>must</b></i> be utilized as part of the returned
      * {@link CapturedContext}, usually wrapped as described above.
-     * @return the wrapped {@link CapturedContext}, or the original if there was no additional state captured.
+     * @return the wrapped {@link CapturedContext}.
      */
     CapturedContext captureContext(CapturedContext underlying);
+
+    /**
+     * Capture a copy of existing context in preparation for an asynchronous thread jump.
```

Comment:
It's not always a thread jump, users can simply wrap something with `defer` to isolate contexts
```suggestion
     * Capture a copy of existing context in preparation for an asynchronous boundary jump.
```